storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1760 @@ buf_block_t *fsp_get_latched_xdes_page(
+    if (err) goto err_exit;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) goto err_exit;
+
+    allocator.complete();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    pageRemover remover(header_block, iblock, inode, extent_size,
+                        cur_page_no, &mtr);
+    err= remover.prepare();
+ 
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      allocator.restore_old_pages();
+      mtr.discard_modifications();
+      goto err_exit;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= allocator.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    mtr.memcpy(*new_block, new_block->page.frame,
+               block->page.frame, srv_page_size - FIL_PAGE_DATA_END);
+    mtr.write<4,mtr_t::FORCED>(*new_block,
+                               FIL_PAGE_OFFSET + new_block->page.frame,
+                               new_page_no);
+
+    /* Assign the new block page number in left, right
+    and parent block */
+    related_pages.complete(new_page_no, parent_offset);
+
+    /* Complete the page free operation */
+    remover.complete(); 
+    /* Add the new page in inode fragment array */
+    allocator.assign_frag_slot();
+
+    if (level)
+    {
+      err= store_child_pages(new_block);
+      if (err) goto err_exit;
+    }
+    goto fetch_next_page;
+  }
+
+  ut_ad(!fsp_tablespace_validate(fil_system.sys_space));
+  ut_ad(!fseg_validate(fil_system.sys_space, m_index));
+  return DB_SUCCESS;
+}
+
+dberr_t indexDefragmenter::defragment(spaceDefragmenter *space_defrag) noexcept
+{
+  mtr_t mtr;
+  dberr_t err= DB_SUCCESS;
+  mtr.start();
+  mtr.x_lock_space(fil_system.sys_space);
+  buf_block_t *root= buf_page_get_gen(page_id_t{0, m_index->page},
+                                      0, RW_SX_LATCH, nullptr,
+                                      BUF_GET_POSSIBLY_FREED, &mtr,
+                                      &err);
+  if (err)
+  {
+    mtr.commit();
+    return err;
+  }
+  uint16_t level= btr_page_get_level(root->page.frame);
+  mtr.commit();
+  while (1)
+  {
+    err= defragment_level(level, space_defrag);
+    DBUG_EXECUTE_IF("fail_after_level_defragment",
+                    if (level == 1)
+		      err= DB_CORRUPTION;);
+    if (err || !level)
+      break;
+    level--;
+  }
+  return err;
+}
+
+static
+bool is_user_table_exists(fil_space_t *space, dberr_t *err) noexcept
+{
+  mtr_t mtr;
+  btr_pcur_t pcur;
+  mtr.start();
+  dict_sys.lock(SRW_LOCK_CALL);
+  for (const rec_t *rec= dict_startscan_system(&pcur, &mtr,
+			                       dict_sys.sys_tables);
+       rec; rec= dict_getnext_system(&pcur, &mtr))
```

Comment:
Do we really need to hold `dict_sys.latch` here? The relevant fields of `dict_sys.sys_tables` (such as the index root page number) should be hard-coded.